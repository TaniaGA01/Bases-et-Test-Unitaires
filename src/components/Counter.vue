<template>
  <h2>{{ CustomTitle }}</h2>
  <p>{{ infoMessage }}</p>
  <h3>Raíz cuadrada usando Method</h3>
  <p>{{ counter }}<sup>2</sup> = {{ RaizCuadradaMethod() }}</p>
  <p data-testId="counter">{{ counter }}</p>
  <div class="button-con">
    <button data-testId="counterBtnInc" @click="IncreaseCounter">+1</button>
    <button data-testId="counterBtnDec" @click="DecreaseCounter">-1</button>
  </div>
  <!-- el @ reemplaza el v:on -->
  <!-- <h3>Raíz cuadrada usando Computed</h3>
  <p>{{ counter }}<sup>2</sup> = {{ RaizCuadradaComputed }}</p> no se usa () con computed -->
</template>

<script>
export default {
  props: {
    // propiedades
    title: String,
    start: {
      type: Number,
      default: 100,
    },
    validator(value) {
      // validación del valor de :start en App.vue
      return value >= 0;
    },
  },
  name: "Contador", // nos permite identificar el componente en el navegador Inspecter -> Vue
  data() {
    return {
      counter: this.start,
    };
  },
  //Methods : se utilizan cuando se necesiten cambiar/alterar los datos, entonces la funcion se ejecutará cada vez que se recargue la página
  methods: {
    RaizCuadradaMethod() {
      return this.counter * this.counter;
    },
    IncreaseCounter() {
      this.counter++;
    },
    DecreaseCounter() {
      this.counter--;
    },
  },
  // Computed : son muy similares a los métodos pero son mejores pq los datos se guardan en cache. Se usan para transformar datos de data en la capa de presentación pero NO deben alterar los datos existentes. Entonces no se ejecuta la función cada vez que se recargue la página ya que solo el dato se actualiza https://vueschool.io/lessons/computed-properties-in-vue-3?friend=vuejs
  computed: {
    RaizCuadradaComputed() {
      console.log("RaizCuadradaComputed");
      return this.counter * this.counter;
    },
    CustomTitle() {
      return this.title || "Counter";
    },
    infoMessage() {
      return (
        this.title || "Comienza en 10 gracias al start en las props (propiedad)"
      );
    },
  },
};
</script>

<style>
</style>