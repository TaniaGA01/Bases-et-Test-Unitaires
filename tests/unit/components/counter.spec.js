/*  
Pour la maj il faut utiliser la commande 'yarn test:unit -u counter'
*/

import { shallowMount} from '@vue/test-utils'
import Counter from '@/components/Counter'

describe('Counter Component', () => {

    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(Counter)
    });
    
    // TEST 1 : SnapShot, une copie du composant sera créée dans un fichier __snapshot__
    test('Avoir match avec le snapshot', () => {

        const wrapper = shallowMount( Counter )

        expect(wrapper.html()).toMatchSnapshot()

    })


    //TEST 2 : HTLM
    test('La valeur par défaut du h2 es "Counter"', () => {
        
        expect(wrapper.find('h2').exists()).toBeTruthy // Si la valeur h2 existe = true
        
        const h2Value = wrapper.find('h2').text()
        
        expect(h2Value).toBe('Counter')
    })


    //TEST 3A : FindAll vs Find

    // test('Dans <p> la valeur par défaut est 100"', () => {

    //     const pValue = wrapper.findAll('p').pop()//dernier de l'array
        
    //     expect(pValue.text()).toBe('100')
    // })


    //TEST 3B : FindAll vs Find

    test('Dans <p> la valeur par défaut est 100"', () => {

        const pValue = wrapper.find('[data-testId="counter"]').text()
        
        expect(pValue).toBe('100')
    })


    //TEST 4A : EVENTS, verification Increase Counter
    test('Augmenteret diminuer en 1 la valeur du Counter', async() => {

        const increaseBtn = wrapper.find('[data-testId="counterBtnInc"]')

        //identifier l'event :
        await increaseBtn.trigger('click')

        //capturer la valeur du p après le click
        const pValue = wrapper.find('[data-testId="counter"]').text()
        
        expect(pValue).toBe('101')

        
    })


    //TEST 4B : EVENTS, verification decrease Counter
    test('Diminuer en 1 la valeur du Counter', async() => {

        const decreaseBtn = wrapper.find('[data-testId="counterBtnDec"]')

        //identifier l'event :
        await decreaseBtn.trigger('click')
        await decreaseBtn.trigger('click')

        //capturer la valeur du p après le double click
        const pValue = wrapper.find('[data-testId="counter"]').text()
        
        expect(pValue).toBe('98')
        
    })

    //TEST 5: Vérification (lire) des props dans Counter.vue
    test('Établir la valeur par défaut', () => {

        const { start } = wrapper.props()
        console.log(`start : `, start)
        console.log(`start type : `, typeof start)

        const pValue = wrapper.find('[data-testId="counter"]').text()

        expect (Number(pValue)).toBe(100)// on transforme le string pValue en nombre
        
    })


    //TEST 6: Affichage de la props 'title'
    test('Afficher la props "title"', () => {

        const title = 'hola mundo !!'

        const wrapper = shallowMount(Counter, { // cette const ignore le let wrapper de la ligne 11
            props: {
                title: title 
            }
        }) 

        expect(wrapper.find('h2').text()).toBe(title)
        
    })



})