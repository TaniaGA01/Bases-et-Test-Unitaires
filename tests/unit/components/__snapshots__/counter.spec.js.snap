// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Avoir match avec le snapshot 1`] = `
<h2>Counter</h2>
<p>Comienza en 10 gracias al start en las props (propiedad)</p>
<h3>Raíz cuadrada usando Method</h3>
<p>100<sup>2</sup> = 10000</p>
<p data-testid="counter">100</p>
<div class="button-con"><button data-testid="counterBtnInc">+1</button><button data-testid="counterBtnDec">-1</button></div><!-- el @ reemplaza el v:on -->
<!-- <h3>Raíz cuadrada usando Computed</h3>
  <p>{{ counter }}<sup>2</sup> = {{ RaizCuadradaComputed }}</p> no se usa () con computed -->
`;
